### PR TITLE
Adding kernbench test to avocado

### PR DESCRIPTION
**Pull request by sudeeshjohn**
Kernbench is a standard CPU benchmark. Runs a kernel
compile and measures the performance by using time
command. By default it runs only once with 2 \* number
of CPUs.

Signed-off-by: Sudeesh John sudeesh@linux.vnet.ibm.com

### CODE REVIEW

File: perf/kernbench.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    def time_build(self, kern,
+                   threads=None, timefile=None, make_opts=None):
+        """time the bulding of the kernel"""
+        os.chdir(kern.linux_dir)
+        build.make(kern.linux_dir, extra_args='clean')
+        if kern.config_path is None:
+            build.make(kern.linux_dir, extra_args='defconfig')
+        else:
+            build.make(kern.linux_dir, extra_args='oldnoconfig')
+        process.system('make silentoldconfig')
```
Comment (PraveenPenguin) [inline perf/kernbench.py]: sorry my bad luck did not notice this , one question why two time kernel make if it is valid why  can  `process.system('make silentoldconfig') `will be `build.make(kern.linux_dir, extra_args='silentoldconfig')` ?

**Comment by PraveenPenguin** [comment]
@sudeeshjohn ping

File: perf/kernbench.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        Kernel build Test
+        """
+        iterations = self.params.get('runs', default=1)
+        threads = self.params.get('cpus', default=None)
+        version = self.params.get('version', default='4.6')
+        kernel_version = platform.uname()[2]
+        config_path = '/boot/config-' + kernel_version
+        if iterations is None:
+            # Setting the default iteration as one
+            iterations = 1
```
Comment (narasimhan-v) [inline perf/kernbench.py]: this is not needed, as [https://github.com/avocado-framework-tests/avocado-misc-tests/pull/35/files#diff-c481269abe19849c8758d2b1522f49c8R92](url) takes care of this

File: perf/kernbench.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        return results
+
+    def test(self):
+        """
+        Kernel build Test
+        """
+        iterations = self.params.get('runs', default=1)
+        threads = self.params.get('cpus', default=None)
+        version = self.params.get('version', default='4.6')
+        kernel_version = platform.uname()[2]
```
Comment (narasimhan-v) [inline perf/kernbench.py]: Please move these assignments to setUp()

File: perf/kernbench.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        config_path = '/boot/config-' + kernel_version
+        if iterations is None:
+            # Setting the default iteration as one
+            iterations = 1
+        if version is None:
+            # setting the default version 4.6
+            version = '4.6'
```
Comment (narasimhan-v) [inline perf/kernbench.py]: Same as above

File: perf/kernbench.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    """
+    Kernbench comples the kernel source fetched from the kernel.org
+    and compile it as per the configuration file. Performance figures will be
+    shown in the log file.
+    """
+
+    def time_build(self, kern,
+                   threads=None, timefile=None, make_opts=None):
+        """time the bulding of the kernel"""
+        os.chdir(kern.linux_dir)
```
Comment (narasimhan-v) [inline perf/kernbench.py]: `build.make()` takes care of this.

File: perf/kernbench.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            results.append(tuple([self.to_seconds(elt) for elt in result]))
+        return results
+
+    def test(self):
+        """
+        Kernel build Test
+        """
+        iterations = self.params.get('runs', default=1)
+        threads = self.params.get('cpus', default=None)
+        version = self.params.get('version', default='4.6')
```
Comment (narasimhan-v) [inline perf/kernbench.py]: Are both `version` and `kernel_version` not same ? Correct me if I am wrong here

File: perf/kernbench.py (hunk 5 above)
Comment (PraveenPenguin) [inline perf/kernbench.py]: @narasimhan-v we need this function as in this test require capture time taken for building kernel  
@sudeeshjohn now  we already have functionality in avocado API to capture   time of build command as function name is build.run_make() 
https://github.com/avocado-framework/avocado/blob/master/avocado/utils/build.py
code snippet can be : 
```
result = build.run_make(kern.linux_dir) 
print result.duration
```

**Comment by PraveenPenguin** [comment]
@sudeeshjohn ping

**Comment by PraveenPenguin** [comment]
@sudeeshjohn ping , any update here

**Comment by sudeeshjohn** [comment]
@PraveenPenguin , sorry for the late reply. Will look at the comments asap. Thanks!

File: perf/kernbench.py (hunk 1 above)
Comment (sudeeshjohn) [inline perf/kernbench.py]: @PraveenPenguin , that was added for testing something, I'll remove it.

File: perf/kernbench.py (hunk 3 above)
Comment (sudeeshjohn) [inline perf/kernbench.py]: @narasimhan-v , sure will do it

File: perf/kernbench.py (hunk 5 above)
Comment (sudeeshjohn) [inline perf/kernbench.py]: @PraveenPenguin , I'll check make the required change

File: perf/kernbench.py (hunk 6 above)
Comment (sudeeshjohn) [inline perf/kernbench.py]: @narasimhan-v , here the "kernel_version" is the version of the kernel the machine booted with. the other "version " refers to the kernel version the test going to build.

File: perf/kernbench.py (hunk 5 above)
Comment (sudeeshjohn) [inline perf/kernbench.py]: @PraveenPenguin , it looks like result.duration provides only total time taken..I think it is good to show user , system and elapsed times

Comment (sudeeshjohn) [inline perf/kernbench.py]: @PraveenPenguin let me know what you think

File: perf/kernbench.py (hunk 1 above)
Comment (PraveenPenguin) [inline perf/kernbench.py]: sure

File: perf/kernbench.py (hunk 5 above)
Comment (PraveenPenguin) [inline perf/kernbench.py]: @sudeeshjohn sure make sense , it is  just a suggestion

**Comment by PraveenPenguin** [comment]
@sudeeshjohn any update here

File: perf/kernbench.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        for run in range(iterations):
+            log.info("Iteration: %s" % run)
+            self.time_build(kern, threads, timefile, "")
+            # Processing the timefile
+            results = open(timefile).readline().strip()
+            (user, system, elapsed) = self.extract_all_time_results(results)[0]
+            user_time += float(user)
+            system_time += float(system)
+            elapsed_time += float(elapsed)
+        # Reults
```
Comment (balamuruhans) [inline perf/kernbench.py]: typo %s/Reults/Results

File: perf/kernbench.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        if version is None:
+            # setting the default version 4.6
+            version = '4.6'
+        if threads is None:
+            # We will use 2 workers of each type for each CPU detected
+            threads = 2 * multiprocessing.cpu_count()
+        # initialize kernel object
+        kern = kernel.KernelBuild(
+            version, config_path=config_path, work_dir=None, data_dirs=None)
+        # Seeting the default kernel URL
```
Comment (balamuruhans) [inline perf/kernbench.py]: %s/seeting/setting

File: perf/kernbench.py.data/kernbench.yaml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# Kernbench configuration file
+# By default, compile the kernel only once with '2 * number of cups'
+# using the default 'config' file in the data directory. 
+# Valid options in avocado test are bellow.
```
Comment (balamuruhans) [inline perf/kernbench.py.data/kernbench.yaml]: %s/bellow/below

File: perf/kernbench.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+from avocado.utils import process
+from avocado.utils import kernel
+from avocado.utils.software_manager import SoftwareManager
+
+log = logging.getLogger('avocado.test')
+
+
+class KernbenchTest(Test):
+    """
+    Kernbench comples the kernel source fetched from the kernel.org
```
Comment (balamuruhans) [inline perf/kernbench.py]: %s/comples/compiles

**Comment by PraveenPenguin** [comment]
@sudeeshjohn this PR is pending from long time are you want it need to still valid one, please suggest ?

**Comment by PraveenPenguin** [comment]
@sudeeshjohn it is pending from more than 6 month as no update if you still want to peruse this please raise new PR , closing this PR

**Pull request closed without merging**